### PR TITLE
bedmaster reader splits up signals

### LINKED ISSUE #1265: Enhance tensorization to split .mat bundles
## What and why
#1258 has shown us that we cannot assume 1:1 relationship between `.mat` file and patient. So tensorizer needs to be able to split up a `.mat` file 

## Solution(s)
algorithm might be:
1. check the end time of a `.mat` file
1. if it overlaps a transfer in/out time, only take values from after/before the overlap

## Acceptance criteria
hd5 files only contain data from 1 patient

### PR DESCRIPTION
**Pull request by StevenSong**
## Issues closed
Closes #1265 

## Changelog
- [x] reader splits signals based on patient transfer in/out times
- [x] some more evidence that I think it's now working in comments down below

## Checklist
- [x] The following are complete:
    - Added or revised tests if appropriate.
    - If new packages are installed, `requirements.txt` is updated.
    - If documentation should be updated, a new issue was made in `ml4c3-wiki`.
    - The PR name describes its content and does not contain "Ref #".
    - The PR does not add PHI.

### CODE REVIEW

**Comment by StevenSong** [comment]
the plot in orange shows the same discontinuity of the `.mat` bundle from #1258 (`LUN06_620-1586396113_7_v4.mat`), where we know the data after the jump belongs to a new patient. this is highlighted by the horizontal gray line which indicates the patient xfer out time from this room

the plot in blue shows the contents of the hd5 generated after the fix in this PR. we see that at the xfer out time, the data from the hd5 continues but it is not the same set of timestamps as that of the `.mat`
image

File: tensorize/bedmaster/readers.py [hunk 1]
```diff
@@ -483,41 +506,65 @@ def get_wv(
...
+        left_idx = 0
+        if left_t is not None:
+            if len(after_left := np.where(time >= left_t)[0]) == 0:
+                # if there are no time points after the left cutoff, no signal
+                return None
+            else:
+                left_idx = after_left[0]
+        right_idx = -1
+        if right_t is not None:
+            if len(before_right := np.where(time <= right_t)[0]) == 0:
```
Comment (RidwanAlam) [inline tensorize/bedmaster/readers.py:520]: do we need a check for invalid timestamps here?

**Review by RidwanAlam — COMMENTED** [review]
Thanks, again @StevenSong

**Comment by RidwanAlam** [comment]
@epalto your review might be more valuable here :)

File: tensorize/bedmaster/readers.py [hunk 2]
```diff
@@ -344,8 +349,26 @@ def get_vs(self, signal_name: str) -> Optional[BedmasterSignal]:
                 "was not found.",
             )
 
-        # Get values and time
-        values = self["vs"][signal_name][()]
+        # Get time and indices for cutoff
+        time = self["vs_time_corrected"][signal_name]["res_vs"][:, 0]
```
[the remark this answers is not in the document]
Comment (StevenSong) [reply tensorize/bedmaster/readers.py:353]: the time array will have `-1` however all the places where we compare the time array, we consider all elements of the array and so the first (few) elements being `-1` should not affect it - looking through some more specific cases where the time vector is used, I dont think it's an issue but perhaps @epalto can double check

File: tensorize/bedmaster/readers.py (hunk 1 above)
Comment (StevenSong) [reply tensorize/bedmaster/readers.py:520]: if invalid means values like `-1`, no, see above

if invalid means values like `nan`, no, it logs a warning but the logic is still correct
image

Comment (RidwanAlam) [reply tensorize/bedmaster/readers.py:520]: cool!

**Review by RidwanAlam — APPROVED** [review]
@StevenSong please wait for @epalto 's review as well. Thanks!

File: tensorize/bedmaster/readers.py [hunk 3]
```diff
@@ -483,41 +506,65 @@ def get_wv(
...
+            else:
+                left_idx = after_left[0]
+        right_idx = -1
+        if right_t is not None:
+            if len(before_right := np.where(time <= right_t)[0]) == 0:
+                # if there are no time points before the right cutoff, no signal
+                return None
+            else:
+                right_idx = before_right[-1] + 1
+        time = time[left_idx:right_idx]
```
Comment (epalto) [inline tensorize/bedmaster/readers.py:525]: If `right_t` is None, then `time = time[0:-1]` which will crop the last value?

File: tensorize/bedmaster/readers.py [hunk 4]
```diff
@@ -628,8 +675,13 @@ def get_wv_from_channel(self, channel: str) -> Optional[str]:
...
 
-    def get_sample_freq_from_channel(self, channel: str, first_idx=0):
-        sf_arr = self["wv_time_original"][channel]["SampleRate"][first_idx:].T[0]
+    def get_sample_freq_from_channel(
+        self,
+        channel: str,
+        first_idx: int = 0,
+        last_idx: int = -1,
+    ):
+        sf_arr = self["wv_time_original"][channel]["SampleRate"][first_idx:last_idx, 0]
```
Comment (epalto) [inline tensorize/bedmaster/readers.py:684]: same than above; [0:-1] will crop the last value

**Review by epalto — CHANGES_REQUESTED** [review]
This looks great! Just a small remark, good job!

File: ingest/icu/tensorizer.py [hunk 5]
```diff
@@ -119,14 +119,16 @@ def tensorize(
...
+        # with multiprocessing.Pool(processes=num_workers) as pool:
+        #     pool.starmap(
+        #         self._main_write,
+        #         [
+        #             (tensors, mrn, visits, ICU_SCALE_UNITS)
+        #             for mrn, visits in files_per_mrn.items()
+        #         ],
+        #     )
+        for mrn, visits in files_per_mrn.items():
+            self._main_write(tensors, mrn, visits, ICU_SCALE_UNITS)
```
[the remark this answers is not in the document]
Comment (StevenSong) [reply ingest/icu/tensorizer.py:131]: oh sorry this was just to test, will restore for sure

File: tensorize/bedmaster/readers.py (hunk 3 above)
Comment (StevenSong) [reply tensorize/bedmaster/readers.py:525]: you're right, will fix

**Comment by StevenSong** [comment]
@epalto any thoughts on how this affects inter-bundle correction? I think it's good but I wanted to get your take on that as well

**Review by RidwanAlam — APPROVED** [review]

**Comment by epalto** [comment]
> @epalto any thoughts on how this affects inter-bundle correction? I think it's good but I wanted to get your take on that as well

I also think that it shouldn't affect it

**Review by epalto — APPROVED** [review]

**Comment by erikr** [comment]
Great work @aguirre-lab/ml4c3-dev 🎉 !!!

**Pull request merged**